### PR TITLE
patch cypher builder, include ListIndex

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Description
Introduce List index, add support to the square bracket notation, updated the docs.

## Complexity

Complexity: Low complexity.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [x] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
┌──────────────────────────────────────────────────┬─────────┬──────────┬──────────────────────────────┬───────────┐
│                     (index)                      │ maxRows │  dbHits  │            cache             │ time (ms) │
├──────────────────────────────────────────────────┼─────────┼──────────┼──────────────────────────────┼───────────┤
│          [33maggregations[0m.TopLevelAggregate          │  1134   │   3403   │   { hits: 41, misses: 0 }    │    62     │
│          [33maggregations[0m.NestedAggregation          │  1134   │   3403   │  { hits: 5905, misses: 0 }   │    324    │
│         [33maggregations[0m.AggregationWithWhere        │  2174   │  13210   │  { hits: 2099, misses: 0 }   │    85     │
│     [33maggregations[0m.NestedCountFromMovieToActors    │  2174   │   8694   │  { hits: 2116, misses: 0 }   │    76     │
│     [33maggregations[0m.NestedCountFromActorsToMovie    │  2174   │   8900   │  { hits: 1613, misses: 0 }   │    70     │
│          [33maggregations[0m.DeeplyNestedCount          │ 2008534 │ 10054408 │ { hits: 5435999, misses: 0 } │   4502    │
│             [33mbatch-create[0m.BatchCreate             │   600   │   3600   │    { hits: 0, misses: 0 }    │    99     │
│             [33mconnect[0m.createAndConnect             │  3003   │  12419   │   { hits: 261, misses: 0 }   │    175    │
│              [33mconnections[0m.Connection              │  2174   │  13042   │  { hits: 3090, misses: 0 }   │    133    │
│           [33mconnections[0m.NestedConnection           │  4516   │  33883   │  { hits: 12961, misses: 0 }  │    118    │
│               [33mcreate[0m.SimpleMutation              │    1    │    6     │    { hits: 0, misses: 0 }    │    41     │
│                [33mdelete[0m.SimpleDelete               │  1040   │  18361   │  { hits: 46366, misses: 0 }  │    285    │
│            [33mdelete[0m.NestedDeleteInUpdate           │  2040   │  16779   │  { hits: 4022, misses: 0 }   │    135    │
│                 [33mquery[0m.SimpleQuery                │  2174   │  14120   │  { hits: 3769, misses: 0 }   │    103    │
│                 [33mquery[0m.QueryWhere                 │  2159   │  12957   │  { hits: 3707, misses: 0 }   │    42     │
│         [33mquery[0m.SimpleQueryWithNestedWhere         │  2159   │  13128   │  { hits: 3748, misses: 0 }   │    52     │
│           [33mquery[0m.OrFilterOnRelationships          │  1731   │  50684   │  { hits: 9543, misses: 0 }   │    149    │
│      [33mquery[0m.OrFilterOnRelationshipsAndNested      │  1731   │  38090   │  { hits: 9433, misses: 0 }   │    272    │
│              [33mquery[0m.QueryWithNestedIn             │  1175   │  14524   │  { hits: 3868, misses: 0 }   │    66     │
│            [33mquery[0m.NestedConnectionWhere           │  2174   │  13182   │  { hits: 3765, misses: 0 }   │    90     │
│         [33mquery[0m.DeeplyNestedConnectionWhere        │  2174   │  13149   │  { hits: 3739, misses: 0 }   │    108    │
│     [33mquery[0m.DeeplyNestedWithRelationshipFilters    │  1134   │   6913   │   { hits: 877, misses: 0 }   │    129    │
│  [33mquery[0m.DeeplyNestedWithRelationshipSingleFilters │  2174   │  15438   │  { hits: 3448, misses: 0 }   │    122    │
│                  [33mquery[0m.Fulltext                  │   16    │    96    │    { hits: 0, misses: 0 }    │    37     │
│           [33mquery[0m.FulltextWithNestedQuery          │   84    │   571    │   { hits: 132, misses: 0 }   │    51     │
│             [33msorting[0m.SortMultipleTypes            │  1040   │   2493   │   { hits: 142, misses: 0 }   │    86     │
│   [33msorting[0m.SortMultipleTypesWithCypherWithCypher  │  1040   │   1449   │  { hits: 3326, misses: 0 }   │    250    │
│            [33msorting[0m.SortOnNestedFields            │  2174   │  13042   │  { hits: 3715, misses: 0 }   │    94     │
│          [33msorting[0m.SortDeeplyNestedFields          │  4516   │  38965   │  { hits: 14495, misses: 0 }  │    207    │
│              [33munions[0m.SimpleUnionQuery             │   35    │   321    │   { hits: 64, misses: 0 }    │    69     │
│     [33munions[0m.SimpleUnionQueryWithMissingFields     │   35    │   293    │   { hits: 54, misses: 0 }    │    57     │
│                [33munions[0m.NestedUnion                │  33033  │  398553  │ { hits: 113399, misses: 0 }  │    306    │
│        [33munions[0m.NestedUnionWithMissingFields       │  33033  │  372527  │ { hits: 107628, misses: 0 }  │    297    │
└──────────────────────────────────────────────────┴─────────┴──────────┴──────────────────────────────┴───────────┘

File: packages/cypher-builder/src/expressions/list/ListExpr.ts [hunk 1]
```diff
@@ -40,7 +49,15 @@ export class ListExpr implements CypherCompilable {
...
     }
 
+    /**
+     * @hidden
+     */
     public getCypher(env: CypherEnvironment): string {
         return this.serializeList(env, this.value);
     }
+
+    /* Access individual elements in the list via the ListIndex class*/
```
Comment (angrykoala) [inline packages/cypher-builder/src/expressions/list/ListExpr.ts:59]: Remember to use ** at the beginning to make it typedocs compatible

```suggestion
    /** Access individual elements in the list via the ListIndex class*/
```

File: packages/cypher-builder/src/expressions/list/ListIndex.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * ```cypher
+ * [ "1", "2", "3" ][0]
+ * ```
+ *
+ */
+export class ListIndex implements CypherCompilable {
+    private value: Variable | ListExpr;
+    private index: number;
+
+    constructor(variable: Variable | ListExpr, index: number) {
```
Comment (angrykoala) [inline packages/cypher-builder/src/expressions/list/ListIndex.ts:43]: This constructor should be hidden, as the user does not have access to this class

File: packages/cypher-builder/src/expressions/list/ListIndex.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+import type { CypherEnvironment } from "../../Environment";
+import type { Variable } from "../../references/Variable";
+import type { CypherCompilable } from "../../types";
+import type { ListExpr } from "./ListExpr";
+
+/** Access individual elements in the list
+ * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/lists/)
+ * @group Expressions
```
Comment (angrykoala) [inline packages/cypher-builder/src/expressions/list/ListIndex.ts:27]: This class is not exported, so it doesn't really makes much sense to document it tbh :sweat_smile:

File: packages/cypher-builder/src/references/PropertyRef.ts [hunk 4]
```diff
@@ -39,10 +39,14 @@ export class PropertyRef implements CypherCompilable {
         return this._variable;
     }
 
+    /* Access individual property via the PropertyRef class, using the dot notation */
```
Comment (angrykoala) [inline packages/cypher-builder/src/references/PropertyRef.ts:42]:
```suggestion
    /** Access individual property via the PropertyRef class, using the dot notation */
```

File: packages/cypher-builder/src/references/Reference.ts [hunk 5]
```diff
@@ -37,9 +38,15 @@ export abstract class Reference implements CypherCompilable {
         return `${id}`;
     }
 
+    /* Access individual property via the PropertyRef class, using the dot notation */
```
Comment (angrykoala) [inline packages/cypher-builder/src/references/Reference.ts:41]:
```suggestion
    /** Access individual property via the PropertyRef class, using the dot notation */
```

**Review by angrykoala — CHANGES_REQUESTED** [review]
Some minor changes in documentation, the code itself looks good

**Pull request merged**